docs/rfc/007-endpoint-errors.md

```diff
@@ -0,0 +1,229 @@
+# [RFC] Conjure Endpoint Errors
+
+**Date:** 2020/03/19
+
+**Stakeholders:**
+* *Anyone that produces or consumes conjure-generated code.*
+
+* * *
+
+## **Context**
+
+Currently, errors can be defined in Conjure type definitions. These errors can then be thrown by services, and if a client 
+knows about them, they can be handled during runtime by clients. The exact handling of this mechanism may vary by
+the implementation (for example, in Java these are `ServicException`s, which become `RemoteException`s which can be caught
+by the client).
+
+There are multiple issues with this. As a consumer, you only know that an endpoint may throw an error by reading the 
+documentation or the implementation itself. As a producer, you are limited in the expressivity of what you can say 
+with your exception, as you have no guarantee that a consumer will handle your exception (or even read your documentation).
```

Comment:
as a bystander here, I'm interested to hear a bit more about the context/motivation here... were there some specific internal cases that motivated this? or is it a broader hope that by defining a new first-class way of declaring errors in conjure that people will use them more?

Reply:
There are some examples that have been experienced in the wild. As an example:

- Consumers of an API assume there are no limits to what can be returned from a batch endpoint.
- They write code that interacts with this.
- At some point they reach a sufficient scale where the batch endpoint hits the limit, and it starts erroring.
- This can result in parts of a UI or service stopping working, and bad times for everyone.

If that endpoint had had a way to say 'hey, btw there is a very real chance of a bad edge case occurring' then they might have programmed defensively against that (of course, in this case they probably should have used pages or something similar too).

I think the idea behind declared errors is to have a generalised way to embed bad edge cases into your API definition.

Reply:
For any other readers I think the specific RFC we're talking about here is titled `[RFC] Handling limits of batch endpoints [2021-03-01]`. I am **very concerned** about the idea of pushing for a `LimitExceededError` to solve this specific case, as I worry it would lead to every client re-implementing the same crap to detect limit exceeded and then do some funky defensive "try again with fewer ids" stuff.

```
batchGetEmployees:
  http: POST ...
  args:
    pageSize:
      param-type: query
      type: optional<string>
    pageToken:
      param-type: query
      type: string
    request: BatchGetEmployeesRequest
  returns: BatchGetEmployeesResponse
  errors:
    - LimitExceededError # <- proposed addition
```

My preferred solution for that RFC is already mentioned lower down: `Implement batching via paging patterns` . This is the way that all other servers I know of handle this, and crucially it can be done transparently on the server side so that all clients benefit without having to write/maintain any funky code.

I think my request here is for a few more _concrete, idiomatic cases_ where this feature would be immediately adopted, because if it's all motivated by this LimitExceeded thing and hypotheticals, then I don't think we should do it.

Reply:
Is there an RFC describing the paging design? I'm concerned about the complexity of implementing such a thing in a general way throughout conjure.

Reply:
The `Implement batching via paging patterns` as I've seen it done across many services doesn't need any additions to conjure - it's just a convention where people define an endpoint with a `pageToken` optional query param, and then return a `FooPage` which contains an optional `nextPageToken` (the contents of which should remain opaque).

In this specific case, i think the solution would be if the user asked for 3000 ids and the limit for backend requests/responses is 1000, then the first page can just return 1000 and include the pagetoken to say there's more to come.

Reply:
While easy to misuse, that solution can work for simple batch loads, but it requires users to verify the result otherwise they may fail silently, assuming all results have been read. Unfortunately it doesn't solve batch store operations where the client has sent a full request, and doesn't necessarily expect a response (or expects an atomic operation).